### PR TITLE
[TS Migration] Migrate EReceiptThumbnail.js component to TypeScript

### PR DESCRIPTION
**Pull request by JKobrynski**
### Fixed Issues

$ https://github.com/Expensify/App/issues/31956
PROPOSAL: N/A

### Tests

(By @grgia)

I think the best way to QA this in DEV is to test the component in the storybook build and look at the EReceiptThumbnail, EReceipt, and ReportPreview stories.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

(by @grgia)

As for production QA tests that cover the EReceiptThumbnail,
We can:
1 . Create a report with 1 - 3 Expensify Card Transactions
- Note: you can do this in OldDot or NewDot. If done in OldDot, use the reportID from the URL and paste it into NewDot to access the report r/<reportID>
2. Ensure that the ReportPreview in the parent chat displays the 1-3 EReceipt Previews ✅
3. Click on the ReportPreview to open the Expense Report.
4. Ensure that each expense displays the correct MCC logo ✅
5. Click on each expense -> Opens MoneyRequest Page -> Click on EReceipt Preview -> Opens EReceipt Page
6. Ensure that the color and MCC logo is consistent on all pages ✅

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

android

</details>

<details>
<summary>Android: mWeb Chrome</summary>

chrome

</details>

<details>
<summary>iOS: Native</summary>

ios

</details>

<details>
<summary>iOS: mWeb Safari</summary>

safari

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

web

</details>

<details>
<summary>MacOS: Desktop</summary>

desktop

</details>

### CODE REVIEW

File: src/components/EReceiptThumbnail.tsx [hunk 1]
```diff
@@ -1,31 +1,28 @@
...
 import Image from './Image';
-import transactionPropTypes from './transactionPropTypes';
-
-const propTypes = {
-    /* TransactionID of the transaction this EReceipt corresponds to */
-    // eslint-disable-next-line react/no-unused-prop-types
-    transactionID: PropTypes.string.isRequired,
 
+type EReceiptThumbnailOnyxProps = {
     /* Onyx Props */
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:18]:
```suggestion
```

File: src/components/EReceiptThumbnail.tsx [hunk 2]
```diff
@@ -114,11 +126,10 @@ function EReceiptThumbnail({transaction}) {
...
 
 EReceiptThumbnail.displayName = 'EReceiptThumbnail';
-EReceiptThumbnail.propTypes = propTypes;
-EReceiptThumbnail.defaultProps = defaultProps;
 
-export default withOnyx({
+export default withOnyx<EReceiptThumbnailProps, EReceiptThumbnailOnyxProps>({
     transaction: {
         key: ({transactionID}) => `${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`,
+        selector: (transaction: OnyxEntry<Transaction>) => transaction,
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:133]:
```suggestion
```

File: src/components/EReceiptThumbnail.tsx [hunk 3]
```diff
@@ -1,31 +1,28 @@
...
     /* Onyx Props */
-    transaction: transactionPropTypes,
+    transaction: OnyxEntry<Transaction>;
 };
 
-const defaultProps = {
-    transaction: {},
+type EReceiptThumbnailProps = EReceiptThumbnailOnyxProps & {
+    /* TransactionID of the transaction this EReceipt corresponds to */
+    // eslint-disable-next-line react/no-unused-prop-types
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:24]: Let's add a comment that it's used in withOnyx HOC

File: src/components/EReceiptThumbnail.tsx [hunk 4]
```diff
@@ -37,30 +34,47 @@ const backgroundImages = {
...
-    const {backgroundColor: primaryColor, color: secondaryColor} = StyleUtils.getEReceiptColorStyles(StyleUtils.getEReceiptColorCode(transaction));
 
     const [containerWidth, setContainerWidth] = useState(0);
     const [containerHeight, setContainerHeight] = useState(0);
 
-    const onContainerLayout = (event) => {
+    const backgroundImage = useMemo(() => {
+        if (!transaction) {
+            return;
+        }
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:47]: Please make sure this update doesn't cause regression. Also maybe consider updating `getEReceiptColorCode` typing

File: src/components/EReceiptThumbnail.tsx [hunk 5]
```diff
@@ -37,30 +34,47 @@ const backgroundImages = {
...
+
+        return backgroundImages[StyleUtils.getEReceiptColorCode(transaction)];
+    }, [StyleUtils, transaction]);
+
+    if (!transaction) {
+        return;
+    }
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:54]: Why this was added? Due to `getEReceiptColorCode` function? Maybe it's better to update its typing, wdyt?

File: src/components/EReceiptThumbnail.tsx [hunk 6]
```diff
@@ -37,30 +34,47 @@ const backgroundImages = {
...
+
+    // Get receipt colorway, or default to Yellow.
+    // const {backgroundColor: primaryColor, color: secondaryColor} = StyleUtils.getEReceiptColorStyles(StyleUtils.getEReceiptColorCode(transaction));
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:57]:
```suggestion
```

File: src/components/EReceiptThumbnail.tsx [hunk 7]
```diff
@@ -1,31 +1,28 @@
...
-import {withOnyx} from 'react-native-onyx';
+import {LayoutChangeEvent, View} from 'react-native';
+import {OnyxEntry, withOnyx} from 'react-native-onyx';
 import * as ReportUtils from '@libs/ReportUtils';
 import useStyleUtils from '@styles/useStyleUtils';
 import useThemeStyles from '@styles/useThemeStyles';
 import variables from '@styles/variables';
 import CONST from '@src/CONST';
 import ONYXKEYS from '@src/ONYXKEYS';
+import {Transaction} from '@src/types/onyx';
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:10]:
```suggestion
import type {Transaction} from '@src/types/onyx';
```

**Review by VickyStash — CHANGES_REQUESTED** [review]

File: src/components/EReceiptThumbnail.tsx (hunk 5 above)
Comment (JKobrynski) [reply src/components/EReceiptThumbnail.tsx]: I updated it, great idea!

File: src/components/EReceiptThumbnail.tsx (hunk 4 above)
Comment (JKobrynski) [reply src/components/EReceiptThumbnail.tsx:48]: Updated typings

File: src/components/EReceiptThumbnail.tsx [hunk 8]
```diff
@@ -37,30 +35,42 @@ const backgroundImages = {
...
     const [containerHeight, setContainerHeight] = useState(0);
 
-    const onContainerLayout = (event) => {
+    const backgroundImage = useMemo(() => {
+        if (!transaction) {
+            return;
+        }
+
+        return backgroundImages[StyleUtils.getEReceiptColorCode(transaction)];
+    }, [StyleUtils, transaction]);
```
Comment (VickyStash) [inline src/components/EReceiptThumbnail.tsx:51]: You can turn it back as it was before
```suggestion
const backgroundImage = useMemo(() => backgroundImages[StyleUtils.getEReceiptColorCode(transaction)], [StyleUtils, transaction]);
```

File: src/styles/utils/index.ts [hunk 9]
```diff
@@ -298,9 +298,9 @@ function getDefaultWorkspaceAvatarColor(workspaceName: string): ViewStyle {
 /**
  * Helper method to return eReceipt color code
  */
-function getEReceiptColorCode(transaction: Transaction): EReceiptColorName {
+function getEReceiptColorCode(transaction: Transaction | null): EReceiptColorName {
```
Comment (VickyStash) [inline src/styles/utils/index.ts:301]:
```suggestion
function getEReceiptColorCode(transaction: OnyxEntry<Transaction>): EReceiptColorName {
```

File: src/components/EReceiptThumbnail.tsx (hunk 8 above)
Comment (JKobrynski) [reply src/components/EReceiptThumbnail.tsx]: Right, updated

File: src/styles/utils/index.ts (hunk 9 above)
Comment (JKobrynski) [reply src/styles/utils/index.ts]: Done!

**Review by VickyStash — APPROVED** [review]
LGTM

**Review by blazejkustra — APPROVED** [review]

**Comment by JKobrynski** [comment]
@0xmiroslav checklist fixed!

**Review by 0xmiroslav — APPROVED** [review]
LGTM!

**Review by 0xmiroslav — APPROVED** [review]

File: src/components/EReceiptThumbnail.tsx [hunk 10]
```diff
@@ -1,31 +1,29 @@
...
+type EReceiptThumbnailOnyxProps = {
+    transaction: OnyxEntry<Transaction>;
 };
 
-const defaultProps = {
-    transaction: {},
+type EReceiptThumbnailProps = EReceiptThumbnailOnyxProps & {
+    /** TransactionID of the transaction this EReceipt corresponds to
+     * It's used by withOnyx HOC
+     */
```
Comment (tgolen) [inline src/components/EReceiptThumbnail.tsx:24]:
```suggestion
    /** TransactionID used by withOnyx of the transaction this EReceipt corresponds to */
```

Prop comments are preferred to be single lines.

File: src/components/EReceiptThumbnail.tsx [hunk 11]
```diff
@@ -37,30 +35,36 @@ const backgroundImages = {
...
-    // Get receipt colorway, or default to Yellow.
-    const {backgroundColor: primaryColor, color: secondaryColor} = StyleUtils.getEReceiptColorStyles(StyleUtils.getEReceiptColorCode(transaction));
 
     const [containerWidth, setContainerWidth] = useState(0);
     const [containerHeight, setContainerHeight] = useState(0);
 
-    const onContainerLayout = (event) => {
+    const backgroundImage = useMemo(() => backgroundImages[StyleUtils.getEReceiptColorCode(transaction)], [StyleUtils, transaction]);
+
+    // Get receipt colorway, or default to Yellow.
```
Comment (tgolen) [inline src/components/EReceiptThumbnail.tsx:47]:
```suggestion
```

I think this comment can be removed, it is not valuable and I don't think it's accurate anymore.

File: src/components/EReceiptThumbnail.tsx [hunk 12]
```diff
@@ -37,30 +35,36 @@ const backgroundImages = {
...
+    const colorStyles = StyleUtils.getEReceiptColorStyles(StyleUtils.getEReceiptColorCode(transaction));
```
Comment (tgolen) [inline src/components/EReceiptThumbnail.tsx:48]: Why don't these consts (and many others in this file) have type declarations?

**Review by tgolen — CHANGES_REQUESTED** [review]
I am learning a lot about TS through these migrations!

**Comment by tgolen** [comment]
Oh, also, please do add QA tests. I would suggest looking through the git blame to find when the feature was implemented and see how it was originally QAed. That might give you some insight into how to test it.

File: src/components/EReceiptThumbnail.tsx (hunk 12 above)
Comment (blazejkustra) [reply src/components/EReceiptThumbnail.tsx:48]: Typescript is automatically inferring it from the the return type of `getEReceiptColorStyles` function:

image

**Comment by blazejkustra** [comment]
@JKobrynski Bump for once you are back 😄

File: src/components/EReceiptThumbnail.tsx (hunk 10 above)
Comment (JKobrynski) [reply src/components/EReceiptThumbnail.tsx]: Done!

File: src/components/EReceiptThumbnail.tsx (hunk 11 above)
Comment (JKobrynski) [reply src/components/EReceiptThumbnail.tsx]: Updated

**Comment by JKobrynski** [comment]
@tgolen this is [the PR](https://github.com/Expensify/App/pull/29075) that introduced this component, it also describes quite a tricky way of testing it 😅 do you think we should do the same for this one?

**Comment by tgolen** [comment]
Thanks for finding that. I am pretty worried that we have a flow that isn't covered by any QA tests or regression tests. There would be no way for us to know whether this is working or not until we hear about it from our customers. @grgia Could you please help write up some QA tests that Applause can perform for testing e-receipts?

**Comment by grgia** [comment]
@tgolen 
I think the best way to QA this in DEV is to test the component in the storybook build and look at the EReceiptThumbnail, EReceipt, and ReportPreview stories.

As for production QA tests that cover the EReceiptThumbnail,
We can:
1 . Create a report with 1 - 3 Expensify Card Transactions
     - **Note:** you can do this in OldDot or NewDot. If done in OldDot, use the reportID from the URL and paste it into NewDot to access the report `r/<reportID>`
2. Ensure that the ReportPreview in the parent chat displays the 1-3 EReceipt Previews ✅ 
3. Click on the ReportPreview to open the Expense Report. 
4. Ensure that each expense displays the correct MCC logo ✅ 
5. Click on each expense -> Opens MoneyRequest Page -> Click on EReceipt Preview -> Opens EReceipt Page
6.  Ensure that the color and MCC logo is consistent on all pages ✅

**Comment by tgolen** [comment]
Thanks, Georgia! @JKobrynski Can you work those into the QA tests, please? Also, it looks like you picked up a conflict.

**Comment by JKobrynski** [comment]
@tgolen updated, conflicts resolved!

**Comment by tgolen** [comment]
Thanks for adding the QA tests. Can you please clean them up a little?

- Put the local testing steps in the "Test" section
- Put the production testing steps in the "QA Steps" section

**Comment by JKobrynski** [comment]
@tgolen done!

**Review by tgolen — APPROVED** [review]

**Pull request merged**